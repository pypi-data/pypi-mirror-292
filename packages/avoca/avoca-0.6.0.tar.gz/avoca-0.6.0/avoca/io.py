from pathlib import Path

import pandas as pd


def to_csv(df: pd.DataFrame, path: Path, **kwargs) -> None:
    """Export a dataframe to a csv file."""
    df.to_csv(path, index=False, **kwargs)


def from_csv(path: Path) -> pd.DataFrame:
    """Read a csv file to a dataframe."""
    df = pd.read_csv(path, header=[0, 1])
    parse_dates = [("-", "datetime"), ("-", "start_datetime"), ("-", "end_datetime")]
    # Convert the datetime column to a datetime object
    for col in parse_dates:
        if col not in df.columns:
            continue
        df[col] = pd.to_datetime(df[col])
    return df
