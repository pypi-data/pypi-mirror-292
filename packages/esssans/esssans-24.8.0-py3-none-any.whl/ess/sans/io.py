# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)

import scipp as sc
import scippnexus as snx
from scippnexus.application_definitions import nxcansas

from .types import BackgroundSubtractedIofQ, OutFilename, RunNumber, RunTitle


def save_background_subtracted_iofq(
    *,
    iofq: BackgroundSubtractedIofQ,
    out_filename: OutFilename,
    run_number: RunNumber,
    run_title: RunTitle,
) -> None:
    """Save background-subtracted I(Q) histogram as an NXcanSAS file."""
    if iofq.bins is None:
        da = iofq.copy(deep=False)
    else:
        da = iofq.hist()
    if da.coords.is_edges('Q'):
        da.coords['Q'] = sc.midpoints(da.coords['Q'])
    with snx.File(out_filename, 'w') as f:
        f['sasentry'] = nxcansas.SASentry(title=run_title, run=run_number)
        f['sasentry']['sasdata'] = nxcansas.SASdata(da, Q_variances='resolutions')
