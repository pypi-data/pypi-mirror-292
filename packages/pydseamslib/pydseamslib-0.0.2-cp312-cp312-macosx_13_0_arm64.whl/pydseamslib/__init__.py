from pydseamslib import cyoda

__all__ = ["cyoda"]
