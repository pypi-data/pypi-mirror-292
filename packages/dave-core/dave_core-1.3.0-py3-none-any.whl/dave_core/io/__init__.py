# Copyright (c) 2022-2024 by Fraunhofer Institute for Energy Economics and Energy System Technology (IEE)
# Kassel and individual contributors (see AUTHORS file for details). All rights reserved.
# Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.

from dave_core.io.convert_format import *
from dave_core.io.file_io import *
from dave_core.io.io_utils import *
