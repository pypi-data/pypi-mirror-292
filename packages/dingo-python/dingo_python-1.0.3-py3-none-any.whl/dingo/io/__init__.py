from dingo.io.input import InputModel, RawInputModel
from dingo.io.output import SummaryModel, ErrorInfo
