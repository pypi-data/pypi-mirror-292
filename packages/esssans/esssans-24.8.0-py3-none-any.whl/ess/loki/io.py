# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
"""
Loading and merging of LoKI data.
"""

import scipp as sc
from ess.reduce import nexus

from ess.sans.types import (
    DetectorEventData,
    Filename,
    MonitorEventData,
    MonitorType,
    NeXusDetector,
    NeXusDetectorName,
    NeXusMonitor,
    NeXusMonitorName,
    RawSample,
    RawSource,
    RunType,
)


def load_nexus_sample(file_path: Filename[RunType]) -> RawSample[RunType]:
    return RawSample[RunType](nexus.load_sample(file_path))


def dummy_load_sample(file_path: Filename[RunType]) -> RawSample[RunType]:
    return RawSample[RunType](
        sc.DataGroup({'position': sc.vector(value=[0, 0, 0], unit='m')})
    )


def load_nexus_source(file_path: Filename[RunType]) -> RawSource[RunType]:
    return RawSource[RunType](nexus.load_source(file_path))


def load_nexus_detector(
    file_path: Filename[RunType], detector_name: NeXusDetectorName
) -> NeXusDetector[RunType]:
    # Events will be loaded later. Should we set something else as data instead, or
    # use different NeXus definitions to completely bypass the (empty) event load?
    dg = nexus.load_detector(
        file_path=file_path,
        detector_name=detector_name,
        selection={'event_time_zero': slice(0, 0)},
    )
    # The name is required later, e.g., for determining logical detector shape
    dg['detector_name'] = detector_name
    return NeXusDetector[RunType](dg)


def load_nexus_monitor(
    file_path: Filename[RunType], monitor_name: NeXusMonitorName[MonitorType]
) -> NeXusMonitor[RunType, MonitorType]:
    return NeXusMonitor[RunType, MonitorType](
        nexus.load_monitor(
            file_path=file_path,
            monitor_name=monitor_name,
            selection={'event_time_zero': slice(0, 0)},
        )
    )


def load_detector_event_data(
    file_path: Filename[RunType], detector_name: NeXusDetectorName
) -> DetectorEventData[RunType]:
    da = nexus.load_event_data(file_path=file_path, component_name=detector_name)
    return DetectorEventData[RunType](da)


def load_monitor_event_data(
    file_path: Filename[RunType], monitor_name: NeXusMonitorName[MonitorType]
) -> MonitorEventData[RunType, MonitorType]:
    da = nexus.load_event_data(file_path=file_path, component_name=monitor_name)
    return MonitorEventData[RunType, MonitorType](da)


providers = (
    load_nexus_detector,
    load_nexus_monitor,
    load_nexus_sample,
    load_nexus_source,
    load_detector_event_data,
    load_monitor_event_data,
)
"""Providers for loading single files."""
