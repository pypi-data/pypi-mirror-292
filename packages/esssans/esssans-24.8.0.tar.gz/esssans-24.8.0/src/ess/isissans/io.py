# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
"""
File loading functions for ISIS data, NOT using Mantid.
"""

from typing import NewType

import sciline
import scipp as sc

from ess.sans.types import (
    BackgroundRun,
    DirectBeam,
    DirectBeamFilename,
    Filename,
    MaskedDetectorIDs,
    PixelMaskFilename,
    RunType,
    SampleRun,
    TransmissionRun,
)

CalibrationFilename = NewType('CalibrationFilename', str)


def read_xml_detector_masking(filename: PixelMaskFilename) -> MaskedDetectorIDs:
    """Read a pixel mask from an XML file.

    The format is as follows, where the detids are inclusive ranges of detector IDs:

    .. code-block:: xml

        <?xml version="1.0"?>
        <detector-masking>
            <group>
                <detids>1400203-1400218,1401199,1402190-1402223</detids>
            </group>
        </detector-masking>

    Parameters
    ----------
    filename:
        Path to the XML file.
    """
    import xml.etree.ElementTree as ET  # nosec

    tree = ET.parse(filename)  # noqa: S314
    root = tree.getroot()

    masked_detids = []
    for group in root.findall('group'):
        for detids in group.findall('detids'):
            for detid in detids.text.split(','):
                detid = detid.strip()
                if '-' in detid:
                    start, stop = detid.split('-')
                    masked_detids += list(range(int(start), int(stop) + 1))
                else:
                    masked_detids.append(int(detid))

    return MaskedDetectorIDs(
        sc.array(dims=['detector_id'], values=masked_detids, unit=None, dtype='int32')
    )


class LoadedFileContents(sciline.Scope[RunType, sc.DataGroup], sc.DataGroup):
    """Contents of a loaded file."""


def load_tutorial_run(filename: Filename[RunType]) -> LoadedFileContents[RunType]:
    return LoadedFileContents[RunType](sc.io.load_hdf5(filename))


def load_tutorial_direct_beam(filename: DirectBeamFilename) -> DirectBeam:
    return DirectBeam(sc.io.load_hdf5(filename))


def transmission_from_sample_run(
    data: LoadedFileContents[SampleRun],
) -> LoadedFileContents[TransmissionRun[SampleRun]]:
    """
    Use transmission from a sample run, instead of dedicated run.
    """
    return LoadedFileContents[TransmissionRun[SampleRun]](data)


def transmission_from_background_run(
    data: LoadedFileContents[BackgroundRun],
) -> LoadedFileContents[TransmissionRun[BackgroundRun]]:
    """
    Use transmission from a background run, instead of dedicated run.
    """
    return LoadedFileContents[TransmissionRun[BackgroundRun]](data)


providers = (read_xml_detector_masking,)
