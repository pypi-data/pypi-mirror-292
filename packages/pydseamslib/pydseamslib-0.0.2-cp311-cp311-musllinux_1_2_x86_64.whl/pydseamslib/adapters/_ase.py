from pydseamslib.cyoda import PointDouble, PointCloudDouble
import numpy as np
import ase


def map_LAMMPS_IDs_to_atomic_symbols(dict_map: dict(), atms_a: ase.Atoms):
    # Traversing by atom is faster than traversing atoms by key value in the map,
    # since there will be fewer map values than atoms
    # See https://github.com/d-SEAMS/PydSEAMSlib/pull/11#issuecomment-2243986351
    # for benchmarks
    for pt in atms_a:
        if pt.number in dict_map.keys():
            pt.symbol = dict_map[pt.number]
    return atms_a


def swap_key_value(my_dict_a):
    new_dict = {value: key for key, value in my_dict_a.items()}
    return new_dict


def list_pts(atms: ase.Atoms, lmp_a: dict, splicemask: list[bool], molids: list[int]):
    ptslist = []
    inslice_atms = atms[splicemask]
    map_lmp = swap_key_value(lmp_a)
    all_atm_id = np.asarray([x.index + 1 for x in atms])[splicemask]
    for idx, atm in enumerate(inslice_atms):
        pd = PointDouble()
        pd.x = atm.position[0]
        pd.y = atm.position[1]
        pd.z = atm.position[2]
        pd.c_type = map_lmp[atm.symbol]
        pd.inSlice = True  # Assumes that the atom provided are in the Slice.
        pd.atomID = all_atm_id[idx]
        pd.molID = molids[idx]
        ptslist.append(pd)
    return ptslist


def to_pointcloud(
    atms_a: ase.Atoms,
    lmp_a: dict,
    splicemask: list[bool],
    molids: list[int],
    fname,
    currentFrame=[1],
):
    _pcd = PointCloudDouble()
    inslice_atms = atms_a[splicemask]
    # TODO(ruhila): Assumes a rectangular box
    _pcd.box = np.diag(np.asarray(inslice_atms.get_cell()))
    _pcd.nop = len(inslice_atms)
    _pcd.pts = list_pts(atms_a, lmp_a, splicemask, molids)
    _pcd.idIndexMap = make_indexmap(atms_a, splicemask)
    _pcd.currentFrame = 1
    _pcd.boxLow = boxlow(fname)
    return _pcd


def make_indexmap(atms: ase.Atoms, splicemask: list[bool]):
    inslice_atms = atms[splicemask]
    atm_id = np.asarray([x.index + 1 for x in atms])[splicemask]
    ret = dict()
    for idx in range(len(inslice_atms)):
        key = int(atm_id[idx])
        ret[key] = idx
    return ret


def boxlow(fname):
    # Read the contents of the file into a variable
    opfile = open(fname)
    data = opfile.read()
    filelist = []
    for i in range(5, 8):
        parfile = data.splitlines()[i]
        needfile = parfile.split()[0]
        floatfile = float(needfile)
        filelist.append(floatfile)
    # Don't forget to close the file again
    opfile.close()
    return filelist
