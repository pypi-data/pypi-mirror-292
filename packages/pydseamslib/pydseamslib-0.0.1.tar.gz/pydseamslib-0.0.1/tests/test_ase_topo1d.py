import pprint
import shutil
import numpy as np

from approvaltests import verify, verify_file
from approvaltests.namer.default_namer_factory import NamerFactory
from ase.io import read as aseread


from pydseamslib import cyoda
from pydseamslib.adapters import _ase
from pathlib import Path

TRAJ = Path("subprojects/seams-core/input/traj/exampleTraj.lammpstrj")
strTRJ = str(TRAJ.absolute())


def test_nlist():
    # Get the frame
    resCloud = cyoda.readLammpsTrjreduced(
        filename=strTRJ,
        targetFrame=1,
        typeI=2,  # oxygenAtomType
        isSlice=False,
        coordLow=[0, 0, 0],
        coordHigh=[0, 0, 0],
    )

    # Construct a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )
    assert pcd.box == resCloud.box
    assert pcd.nop == resCloud.nop
    assert pcd.idIndexMap == resCloud.idIndexMap
    assert pcd.currentFrame == resCloud.currentFrame
    assert pcd.boxLow == resCloud.boxLow

    for idx in range(len(pcd.pts)):
        assert pcd.pts[idx].x == resCloud.pts[idx].x
        assert pcd.pts[idx].y == resCloud.pts[idx].y
        assert pcd.pts[idx].z == resCloud.pts[idx].z
        assert pcd.pts[idx].c_type == resCloud.pts[idx].c_type
        assert pcd.pts[idx].inSlice == resCloud.pts[idx].inSlice
        assert pcd.pts[idx].atomID == resCloud.pts[idx].atomID
        assert pcd.pts[idx].molID == resCloud.pts[idx].molID

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Validate the results
    verify(pprint.pformat(nl))


def test_hbnlist():
    # Get the frame, make a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )

    # Validate the results
    verify(pprint.pformat(hl))


def test_hbnlist1():
    # Get the frame, make a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )
    # Hydrogen-bonded network using indices not IDs
    hL = cyoda.neighbourListByIndex(
        yCloud=pcd,
        nList=hl,
    )

    # Validate the results
    verify(pprint.pformat(hL))


def test_rings():
    # Get the frame, make a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )
    # Hydrogen-bonded network using indices not IDs
    hL = cyoda.neighbourListByIndex(
        yCloud=pcd,
        nList=hl,
    )
    # Gets every ring (non-primitives included)
    Rgs = cyoda.ringNetwork(
        nList=hL,
        maxDepth=6,
    )

    # Validate the results
    verify(pprint.pformat(Rgs))


def test_prisms():
    # Construct a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )
    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )
    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )
    # Hydrogen-bonded network using indices not IDs
    hL = cyoda.neighbourListByIndex(
        yCloud=pcd,
        nList=hl,
    )
    # Gets every ring (non-primitives included)
    Rgs = cyoda.ringNetwork(
        nList=hL,
        maxDepth=6,
    )
    # Ensure the directory is not present before beginning
    outDir = "runOne/"  # / is important for the C++ engine..
    if Path(outDir).exists():
        shutil.rmtree(outDir)
    # Does the prism analysis for quasi-one-dimensional ice
    cyoda.prismAnalysis(
        path=outDir,
        rings=Rgs,
        nList=hL,
        yCloud=pcd,
        maxDepth=6,
        atomID=0,
        firstFrame=1,  # targetFrame
        currentFrame=1,  # frame
        doShapeMatching=False,
    )
    # Validate the run results
    verify_file(
        Path(f"{outDir}/topoINT/dataFiles/system-prisms-1.data"),
        options=NamerFactory.with_parameters("systemPrisms"),
    )
    verify_file(
        Path(f"{outDir}/topoINT/nPrisms.dat"),
        options=NamerFactory.with_parameters("nPrisms"),
    )
