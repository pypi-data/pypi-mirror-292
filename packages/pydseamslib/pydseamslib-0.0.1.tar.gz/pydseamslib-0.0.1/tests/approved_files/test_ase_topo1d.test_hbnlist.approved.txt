[[1, 154, 244, 433],
 [4, 268, 670, 658],
 [7, 73, 52],
 [10, 322, 346, 292, 706],
 [13, 661, 724],
 [16, 433, 727, 187, 280],
 [19, 61, 91, 34, 223],
 [22, 265, 304, 502],
 [25, 190, 619, 205],
 [28, 703, 406],
 [31, 88, 415, 352, 439],
 [34, 19, 394, 58, 277],
 [37, 424, 484, 82, 166],
 [40, 331, 742, 259, 565],
 [43, 115, 448, 625],
 [46, 214, 223, 91, 535],
 [49, 64, 490, 298, 631],
 [52, 7, 529, 145, 334],
 [55, 553, 571, 121, 388],
 [58, 34, 469, 124, 256],
 [61, 19, 736, 748, 283],
 [64, 49, 385, 502, 718],
 [67, 553, 571, 190, 235],
 [70, 364, 628, 94, 511],
 [73, 7, 664, 709, 334],
 [76, 430, 514, 133, 286],
 [79, 286, 589, 133, 487],
 [82, 37, 493, 352, 439],
 [85, 94, 706, 628],
 [88, 31, 178, 715, 670],
 [91, 19, 46, 394, 283],
 [94, 85, 70, 244, 220],
 [97, 331, 742, 319, 733],
 [100, 565, 733, 331, 676],
 [103, 610, 481],
 [106, 247, 655, 112, 160],
 [109, 184, 271, 310, 574],
 [112, 106, 355, 241, 697],
 [115, 43, 556, 691, 463],
 [118, 547, 592, 262, 403],
 [121, 55, 400, 175, 544],
 [124, 58, 130, 394, 421],
 [127, 217, 211, 442],
 [130, 124, 256, 460, 325],
 [133, 76, 79, 211, 442],
 [136, 391, 544, 466, 712],
 [139, 427, 595, 688],
 [142, 211, 412, 373, 496],
 [145, 52, 529, 616, 679],
 [148, 157, 451, 274, 295],
 [151, 160, 655, 583, 613],
 [154, 1, 163, 181, 580],
 [157, 148, 262, 403, 268],
 [160, 151, 106, 697],
 [163, 154, 550, 244],
 [166, 37, 493, 622, 643],
 [169, 661, 187, 280],
 [172, 253, 685, 409, 490],
 [175, 121, 235, 571, 700],
 [178, 88, 439, 622, 493],
 [181, 154, 433, 550],
 [184, 109, 259, 739, 229],
 [187, 16, 169, 427, 526],
 [190, 25, 67, 466, 391],
 [193, 508, 604, 358],
 [196, 304, 502, 397, 505],
 [199, 436, 712, 448, 682],
 [202, 241, 559, 622, 643],
 [205, 25, 682, 466, 712],
 [208, 514, 286, 667],
 [211, 142, 127, 133, 487],
 [214, 46, 277, 421, 394],
 [217, 127, 487, 499, 589],
 [220, 94, 706, 322, 364],
 [223, 46, 19, 748, 277],
 [226, 328, 559, 241, 355],
 [229, 184, 271, 517, 541],
 [232, 406, 496, 412, 703],
 [235, 175, 67, 466, 544],
 [238, 319, 337, 331, 676],
 [241, 202, 112, 226, 418],
 [244, 94, 1, 163],
 [247, 106, 301, 355, 532],
 [250, 295, 472, 409, 685],
 [253, 172, 343, 634, 592],
 [256, 130, 58, 376, 508],
 [259, 184, 40, 310, 379],
 [262, 157, 118, 472, 295],
 [265, 22, 631, 637],
 [268, 4, 157, 451, 445],
 [271, 109, 229, 307, 700],
 [274, 148, 403, 547],
 [277, 214, 34, 223, 469],
 [280, 16, 169, 580],
 [283, 61, 91, 337, 535],
 [286, 79, 76, 208, 568],
 [289, 448, 361, 436],
 [292, 10, 610, 457, 475],
 [295, 250, 148, 262],
 [298, 49, 718, 343],
 [301, 247, 583, 478],
 [304, 22, 196, 637, 694],
 [307, 271, 541, 574, 400],
 [310, 259, 109, 382, 742],
 [313, 682, 361, 436],
 [316, 358, 367, 601, 604],
 [319, 238, 97, 640, 730],
 [322, 10, 220, 475, 481],
 [325, 130, 367, 376],
 [328, 226, 418, 478, 484],
 [331, 40, 97, 100, 238],
 [334, 52, 73, 406, 607],
 [337, 238, 283, 640, 736],
 [340, 565, 733, 382, 742],
 [343, 253, 298, 649, 490],
 [346, 10, 349, 481, 610],
 [349, 346, 430, 667, 514],
 [352, 31, 82, 658, 715],
 [355, 112, 247, 226, 478],
 [358, 316, 193, 613, 460],
 [361, 289, 313, 463, 556],
 [364, 70, 220, 457],
 [367, 316, 325, 460, 586],
 [370, 625, 679, 691],
 [373, 142, 487, 499],
 [376, 256, 325, 586, 469],
 [379, 259, 739, 382, 565],
 [382, 310, 340, 379, 574],
 [385, 64, 490, 649],
 [388, 55, 400, 517, 541],
 [391, 136, 190, 619, 553],
 [394, 34, 91, 124, 214],
 [397, 196, 694, 745],
 [400, 121, 388, 307, 700],
 [403, 157, 274, 118, 445],
 [406, 232, 28, 334, 529],
 [409, 172, 250, 634],
 [412, 142, 232, 499],
 [415, 31, 646, 658, 670],
 [418, 328, 241, 697],
 [421, 214, 124, 469],
 [424, 37, 439, 622, 559],
 [427, 139, 187, 724, 661],
 [430, 76, 349, 442, 568],
 [433, 16, 181, 1, 511],
 [436, 199, 289, 313, 619],
 [439, 178, 424, 31, 82],
 [442, 430, 127, 133, 589],
 [445, 403, 268, 658, 646],
 [448, 43, 289, 199, 463],
 [451, 148, 268, 646, 670],
 [454, 673, 598, 694],
 [457, 364, 292, 628, 706],
 [460, 130, 367, 358, 508],
 [463, 448, 115, 361, 682],
 [466, 190, 235, 136, 205],
 [469, 58, 421, 277, 376],
 [472, 250, 262, 592, 634],
 [475, 322, 292],
 [478, 328, 355, 301, 697],
 [481, 346, 103, 322, 514],
 [484, 37, 328, 643, 559],
 [487, 217, 373, 79, 211],
 [490, 49, 385, 172, 343],
 [493, 82, 166, 178, 715],
 [496, 232, 142, 529],
 [499, 217, 412, 373, 703],
 [502, 64, 196, 22, 631],
 [505, 196, 745, 631, 637],
 [508, 193, 460, 256, 586],
 [511, 433, 70, 550],
 [514, 76, 208, 349, 481],
 [517, 229, 388, 571, 700],
 [520, 532, 601, 538, 604],
 [523, 562, 598, 745],
 [526, 187, 577, 727],
 [529, 52, 145, 406, 496],
 [532, 520, 247, 583, 655],
 [535, 46, 283, 640, 748],
 [538, 520, 613, 655],
 [541, 307, 229, 388],
 [544, 136, 121, 235, 553],
 [547, 118, 274],
 [550, 163, 181, 511, 628],
 [553, 55, 67, 391, 544],
 [556, 115, 361, 709],
 [559, 202, 226, 424, 484],
 [562, 523, 688, 595],
 [565, 100, 340, 40, 379],
 [568, 430, 286, 589, 667],
 [571, 55, 67, 175, 517],
 [574, 307, 109, 382, 739],
 [577, 526, 580],
 [580, 577, 154, 280, 727],
 [583, 301, 532, 151],
 [586, 376, 508, 367, 604],
 [589, 79, 568, 217, 442],
 [592, 118, 472, 253, 685],
 [595, 139, 562, 598, 661],
 [598, 523, 595, 454, 652],
 [601, 520, 316, 613],
 [604, 193, 586, 316, 520],
 [607, 334, 679, 664],
 [610, 103, 292, 346],
 [613, 358, 601, 151, 538],
 [616, 145, 703],
 [619, 25, 391, 436, 712],
 [622, 178, 424, 166, 202],
 [625, 370, 43, 664, 709],
 [628, 70, 457, 85, 550],
 [631, 265, 502, 49, 505],
 [634, 253, 472, 409, 649],
 [637, 304, 265, 505, 652],
 [640, 319, 337, 535, 721],
 [643, 484, 166, 202],
 [646, 415, 451, 445],
 [649, 343, 634, 385, 718],
 [652, 598, 637, 745, 694],
 [655, 106, 151, 532, 538],
 [658, 415, 445, 4, 352],
 [661, 13, 169, 427, 595],
 [664, 73, 625, 607],
 [667, 349, 208, 568],
 [670, 4, 451, 88, 415],
 [673, 454, 688],
 [676, 100, 238, 730, 736],
 [679, 370, 607, 145],
 [682, 205, 313, 199, 463],
 [685, 172, 592, 250],
 [688, 562, 139, 673, 724],
 [691, 115, 370],
 [694, 304, 397, 454, 652],
 [697, 160, 478, 112, 418],
 [700, 271, 400, 175, 517],
 [703, 28, 499, 232, 616],
 [706, 85, 220, 10, 457],
 [709, 73, 556, 625],
 [712, 199, 619, 136, 205],
 [715, 88, 493, 352],
 [718, 298, 649, 64],
 [721, 640, 730, 736, 748],
 [724, 427, 13, 688],
 [727, 16, 526, 580],
 [730, 721, 319, 676],
 [733, 100, 340, 97],
 [736, 61, 721, 337, 676],
 [739, 184, 379, 574],
 [742, 40, 97, 310, 340],
 [745, 505, 652, 397, 523],
 [748, 61, 223, 535, 721]]
